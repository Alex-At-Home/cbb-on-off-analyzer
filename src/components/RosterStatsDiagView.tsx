// React imports:
import React, { useState } from 'react';

// Next imports:
import { NextPage } from 'next';

// Utils
import { ORtgDiagnostics } from "../utils/StatsUtils";


type Props = {
  ortgDiags: ORtgDiagnostics
};
const RosterStatsDiagView: React.FunctionComponent<Props> = ({ortgDiags}) => {

  const [ showMoreORtgPts, setShowMoreORtgPts ] = useState(false);
  const [ showMoreORtgPoss, setShowMoreORtgPoss ] = useState(false);

  const d = ortgDiags;
  return <span>
    ORtg = Points_Produced/Adjusted_Possessions [<b>{d.ptsProd.toFixed(1)}</b>] / [<b>{d.adjPoss.toFixed(1)}</b>]
    &nbsp;(<a href="#" onClick={(event) => { event.preventDefault(); setShowMoreORtgPts(!showMoreORtgPts) }}>{showMoreORtgPts ? "less" : "more"} about points</a>)
    &nbsp;(<a href="#" onClick={(event) => { event.preventDefault(); setShowMoreORtgPoss(!showMoreORtgPoss) }}>{showMoreORtgPoss ? "less" : "more"} about possessions</a>)
    <ul>
      <li>Points_Produced: [<b>{d.ptsProd.toFixed(1)}</b>] = PProd_From_ORB [<b>{d.ppOrb.toFixed(1)}</b>] +
      [<b>{((d.ppFg + d.ppAssist + d.rawFtm)*(1 - d.teamOrbContribPct)).toFixed(1)}</b>]
      (PProd_From_FG [<b>{d.ppFg.toFixed(1)}</b>] + FTM [<b>{d.rawFtm}</b>] + PProd_From_AST [<b>{d.ppAssist.toFixed(1)}</b>],
      less Team_ORB_Contrib% [<b>{(100*d.teamOrbContribPct).toFixed(1)}%</b>]==[<b>{((d.ppFg + d.ppAssist + d.rawFtm)*d.teamOrbContribPct).toFixed(1)}</b>])
      </li>
      <ul>
        <li><em>Compare raw stats: orb=[<b>{d.rawOrb}</b>], pts=[<b>{d.rawPts}</b>] (fg_pts=[<b>{d.ptsFgm}</b>] + ftm=[<b>{d.rawFtm}</b>]), assists=[<b>{d.rawAssist}</b>]</em></li>
      </ul>
      {showMoreORtgPts ?
      <span><li><u>Points section</u></li>
      <ul>
        <li>PProd_From_FG: [<b>{d.ppFg.toFixed(1)}</b>] = Raw_FG_Pts [<b>{d.ptsFgm}</b>], less Team_Assist_Contrib% [<b>{(100*d.ppFgTeamAstPct).toFixed(1)}%</b>]===[<b>{(d.ptsFgm-d.ppFg).toFixed(1)}</b>]</li>
        <ul>
          <li><em>(Team_Assist_Contrib%: In the same way that Player gets rewarded for assisting their team, they get slightly less rewarded when they scored off an assist):</em></li>
          <li>Team_Assist_Contrib% [<b>{(100*d.ppFgTeamAstPct).toFixed(1)}%</b>] = Weighting ([<b>0.5</b>] * Player_eFG [<b>{(100*d.eFG).toFixed(1)}%</b>]) * Team_Assist_Rate [<b>{(100*d.teamAssistRate).toFixed(1)}%</b>]</li>
          <ul>
            <li><em>(The theory behind the weighting is that easier shots are harder to assist, so the higher the eFG the more credit to the assist.)</em></li>
            <li>Team_Assist_Rate: [<b>{(100*d.teamAssistRate).toFixed(1)}%</b>] = (Weighting [<b>1.14</b>] * (Others_AST [<b>{d.othersAssist}</b>] / Team_FGM [<b>{d.teamFgm}</b>])</li>
          </ul>
        </ul>
        <li>PProd_From_AST: [<b>{d.ppAssist.toFixed(1)}</b>] = Weighting ([<b>0.5</b>] * Team_Not_Player_eFG [<b>{(100*d.otherEfg).toFixed(1)}%</b>]) * Team_Not_Player_Pts_Per_Made_Shot [<b>{d.otherPtsPerFgm.toFixed(1)}</b>] * Player_Assists [<b>{d.rawAssist.toFixed(1)}</b>]</li>
        <li>PProd_From_ORB: [<b>{d.ppOrb.toFixed(1)}</b>] = Team_ORB_Weight [<b>{(100*d.teamOrbWeight).toFixed(1)}%</b>] * ORB [<b>{d.rawOrb.toFixed(1)}</b>] * % Plays_Team_Scored [<b>{(100*d.teamScoredPlayPct).toFixed(1)}%</b>] * Pts_Per_Scoring_Possession [<b>{d.teamPtsPerScore.toFixed(1)}</b>] </li>
        <ul>
          <li><em>(Team_ORB_Weight is described under Team_ORB_Contrib%, below)</em></li>
          <li><em>(% Plays_Team_Scored is described in the possessions section, below)</em></li>
          <li>Pts_Per_Scoring_Possession: [<b>{d.teamPtsPerScore.toFixed(1)}</b>] = Team_Pts [<b>{d.teamPts}</b>] / Team_Scoring_Plays [<b>{d.teamScoringPoss.toFixed(1)}</b>]</li>
          <ul>
            <li><em>(Team_Scoring_Plays described under possessions section, below)</em></li>
          </ul>
        </ul>
        <li><em>(Team_ORB_Contrib%: In the same way a Player gets rewarded for an ORB, they get slightly less reward for a score that comes off someone else's rebound):</em></li>
        <li>Team_ORB_Contrib%: [<b>{(100*d.teamOrbContribPct).toFixed(1)}</b>%] = Team_ORB_Weight [<b>{(100*d.teamOrbWeight).toFixed(1)}%</b>] * % Team_Scoring_Plays_From_Rebound [<b>{(100*d.teamScoreFromReboundPct).toFixed(1)}%</b>]</li>
        <ul>
          <li>% Team_Scoring_Plays_From_Rebound: [<b>{(100*d.teamScoreFromReboundPct).toFixed(1)}%</b>] = (Team_ORB [<b>{d.teamOrb}</b>] * % Plays_Team_Scored [<b>{(100*d.teamScoredPlayPct).toFixed(1)}%</b>]) / Scoring_Plays [<b>{d.teamScoringPoss.toFixed(1)}</b>] </li>
          <ul>
            <li><em>(% Plays_Team_Scored is described in the possessions section, below)</em></li>
          </ul>
          <li>Team_ORB_Weight: [<b>{(100*d.teamOrbWeight).toFixed(1)}%</b>] = % Credit_To_Rebounder [<b>{(100*d.teamOrbCreditToRebounder).toFixed(1)}%</b>] / (
          % Credit_To_Rebounder [<b>{(100*d.teamOrbCreditToRebounder).toFixed(1)}%</b>] + % Credit_To_Scorer [<b>{(100*d.teamOrbCreditToScorer).toFixed(1)}%</b>])</li>
          <ul>
            <li><em>(The theory here is that we assign credit to the rebounder based on the relative difficulty of rebounding vs scoring)</em></li>
            <li>% Credit_To_Rebounder [<b>{(100*d.teamOrbCreditToRebounder).toFixed(1)}%</b>] = Team_No_ORB% [<b>{(100*(1 - d.teamOrbPct)).toFixed(1)}%</b>] * Team_Score% [<b>{(100*d.teamScoredPlayPct).toFixed(1)}%</b>]</li>
            <li>% Credit_To_Scorer = [<b>{(100*d.teamOrbCreditToScorer).toFixed(1)}%</b>] = Team_ORB% [<b>{(100*d.teamOrbPct).toFixed(1)}%</b>] * Team_No_Score% [<b>{(100*(1 - d.teamScoredPlayPct)).toFixed(1)}%</b>]</li>
          </ul>
        </ul>
      </ul></span> : null }
      <li>Adjusted_Possessions: [<b>{d.adjPoss.toFixed(1)}</b>] = Scoring_Possessions [<b>{d.scoringPoss.toFixed(1)}</b>] + Missed_FG_Possessions [<b>{d.fgxPoss.toFixed(1)}</b>] + Missed_FT_Possessions [<b>{d.ftxPoss.toFixed(1)}</b>] + TO [<b>{d.rawTo}</b>]</li>
      <ul>
        <li><em>Compare raw stats: poss=[<b>{d.offPoss.toFixed(1)}</b>] (fga=[<b>{d.rawFga}</b>] + 0.475*fta=[<b>{d.ftPoss.toFixed(1)}</b>] + to=[<b>{d.rawTo}</b>] - orb=[<b>{d.offPlaysLessPoss.toFixed(1)}</b>])</em></li>
      </ul>
      {showMoreORtgPoss ?
      <span><li><u>Possessions section</u></li>
      <ul>
        <li>Scoring_Possessions: [<b>{d.scoringPoss.toFixed(1)}</b>] = ORB_Part [<b>{d.orbPart.toFixed(1)}</b>] + (FG_Part [<b>{d.fgPart.toFixed(1)}</b>] + FT_Part [<b>{d.ftPart.toFixed(1)}</b>] + AST_Part [<b>{d.astPart.toFixed(1)}</b>],
        less Team_ORB_Contrib% [<b>{(100*d.teamOrbContribPct).toFixed(1)}%</b>]==[<b>{((d.fgPart + d.ftPart + d.astPart)*d.teamOrbContribPct).toFixed(1)}</b>])</li>
        <ul>
          <li>ORB_Part: [<b>{d.orbPart.toFixed(1)}</b>] = Team_ORB_Weight [<b>{(100*d.teamOrbWeight).toFixed(1)}%</b>]
          * Scoring_Plays_From_Rebound [<b>{(d.rawOrb*d.teamScoredPlayPct).toFixed(1)}</b>] (ORB [<b>{d.rawOrb}</b>] * % Plays_Team_Scored [<b>{(100*d.teamScoredPlayPct).toFixed(1)}%</b>]) </li>
          <ul>
            <li>% Plays_Team_Scored: [<b>{(100*d.teamScoredPlayPct).toFixed(1)}%</b>] = Team_Scoring_Plays [<b>{d.teamScoringPoss.toFixed(1)}</b>] / Team_Total_Plays [<b>{d.teamPlays.toFixed(1)}</b>]</li>
            <li>Team_Scoring_Plays: [<b>{d.teamScoringPoss.toFixed(1)}</b>] = Team_FGM [<b>{d.teamFgm}</b>] + Team_FTs_Hit_1+ [<b>{d.teamFtHitOnePlus.toFixed(1)}</b>]</li>
            <ul>
              <li>Team_FTs_Hit_1+: [<b>{d.teamFtHitOnePlus.toFixed(1)}</b>] = (1 - Team_FTs_Missed_Both% [<b>{(100*(1 - d.teamProbFtHitOnePlus)).toFixed(1)}%</b>]) * (0.475*Team_FTA) [<b>{(0.475*d.teamFta).toFixed(1)}</b>]</li>
              <ul>
                <li>Team_FTs_Missed_Both%: [<b>{(100*(1 - d.teamProbFtHitOnePlus)).toFixed(1)}%</b>] = (1 - Team_FT% [<b>{(100*d.teamFtPct).toFixed(1)}%</b>])^2</li>
                <li><em>(0.475*FTA is a standard equation for estimating the number of trips to the FT line)</em></li>
              </ul>
            </ul>
            <li>Team_Total_Plays: [<b>{d.teamPlays.toFixed(1)}</b>] = Team_FGA [<b>{d.teamFga}</b>] + (0.475*Team_FTA) [<b>{(0.475*d.teamFta).toFixed(1)}</b>] + Team_TOV [<b>{d.teamTo}</b>]</li>
            <li><em>(Team_ORB_Weight is described in the points section, above. The cost of the play is reduced like the reward of the score)</em></li>
          </ul>
          <li>FG_Part: [<b>{d.fgPart.toFixed(1)}</b>] = FGM [<b>{d.rawFgm}</b>], less Team_Assist_Contrib% [<b>{(100*d.ppFgTeamAstPct).toFixed(1)}%</b>]===[<b>{(d.rawFgm*d.ppFgTeamAstPct).toFixed(1)}</b>]</li>
          <ul>
            <li><em>(Team_Assist_Contrib% is described in the points section, above. Since it reduces the reward of a score, it also reduces the cost of the play)</em></li>
          </ul>
          <li>FT_Part: [<b>{d.ftPart.toFixed(1)}</b>] = (1 - Missed_Both_FTs% [<b>{(100*d.missedBothFTs).toFixed(1)}%</b>]) * 0.475*FTA [<b>{d.ftPoss.toFixed(1)}</b>]</li>
          <ul>
            <li>Missed_Both_FTs%: [<b>{(100*d.missedBothFTs).toFixed(1)}%</b>] = (1 - FT% [<b>{(100*d.ftPct).toFixed(1)}%</b>])^2</li>
          </ul>
          <li>AST_Part: [<b>{d.astPart.toFixed(1)}</b>] = Weighting ([<b>0.5</b>] * Team_Not_Player_eFG [<b>{(100*d.otherEfg).toFixed(1)}%</b>]) * AST [<b>{d.rawAssist}</b>]</li>
          <li><em>(Team_ORB_Contrib% is described in the points section, above. Since it reduces the reward of a score, it also reduces the cost of the play)</em></li>
        </ul>
        <li>Missed_FG_Possessions: [<b>{d.fgxPoss.toFixed(1)}</b>] = Missed_FG [<b>{d.rawFgx.toFixed(1)}</b>], less Team_Rebound_Weight [<b>{(107*d.teamOrbPct).toFixed(1)}%</b>]==[<b>{(1.07*d.teamOrbPct*d.rawFgx).toFixed(1)}</b>]</li>
        <ul>
          <li><em>(Team_Rebound_Weight: each missed shot doesn't count a full possession, because it may be rebounded)</em></li>
          <li>Team_Rebound_Weight: [<b>{(107*d.teamOrbPct).toFixed(1)}%</b>] = Weight [<b>1.07</b>] * Team_ORB% [<b>{(100*d.teamOrbPct).toFixed(1)}%</b>]</li>
        </ul>
        <li>Missed_FT_Possessions: [<b>{d.ftxPoss.toFixed(1)}</b>] = Missed_Both_FTs% [<b>{(100*d.missedBothFTs).toFixed(1)}%</b>] * 0.475*FTA [<b>{d.ftPoss.toFixed(1)}</b>]</li>
        <ul>
          <li><em>(FT possession calcs above, under Team_Scoring_Plays)</em></li>
        </ul>
      </ul></span> : null }
    </ul>
  </span>;
};

export default RosterStatsDiagView;
