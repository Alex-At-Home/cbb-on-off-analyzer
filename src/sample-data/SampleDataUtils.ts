import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
