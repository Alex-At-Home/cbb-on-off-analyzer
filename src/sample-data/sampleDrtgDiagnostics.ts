export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
