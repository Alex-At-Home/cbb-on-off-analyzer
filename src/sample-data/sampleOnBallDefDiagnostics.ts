/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
