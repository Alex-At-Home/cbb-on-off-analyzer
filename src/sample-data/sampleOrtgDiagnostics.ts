/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOrtgDiagnostics =
{
   "rawFga": 356,
   "rawFgx": 217,
   "rawFgm": 139,
   "ptsFgm": 334,
   "rawFtm": 172,
   "rawAssist": 147,
   "rawAssistInfo": [
      "70",
      "10",
      "67"
   ],
   "rawPts": 506,
   "rawOrb": 17,
   "rawTo": 69,
   "raw3Fga": 174,
   "raw2midFga": 65,
   "raw2rimFga": 117,
   "raw3Fgm": 56,
   "raw2midFgm": 18,
   "raw2rimFgm": 65,
   "teamOrb": 310,
   "teamPts": 1937,
   "teamFga": 1550,
   "teamFgm": 643,
   "teamFta": 579,
   "teamFtPct": 0.7564766839378239,
   "teamOrbPct": 0.3150406504065041,
   "teamTo": 299,
   "teamPoss": 1802,
   "eFG": 0.4691011235955056,
   "teamPtsPerScore": 2.1481287816538432,
   "teamFtHitOnePlus": 258.7150259067358,
   "teamProbFtHitOnePlus": 0.9406963945340815,
   "rosterOrb": 0,
   "teamOrbCreditToRebounder": 0.29078666101564715,
   "teamOrbCreditToScorer": 0.18129604370791266,
   "teamScoreFromReboundPct": 0,
   "teamOrbWeight": 0.6159655037265658,
   "othersAssist": 214,
   "otherEfg": 0.48785594639866,
   "otherEfgInfo": [
      "48.8",
      "48.8",
      "48.8"
   ],
   "otherPtsPerFgm": 2.3115079365079363,
   "teamOrbContribPct": 0,
   "teamScoredPlayPct": 0.4245312677142386,
   "teamAssistRate_Classic": 0.3794090202177294,
   "ppFgTeamAstPct_Classic": 0.08899059884320337,
   "teamAssistRate": 0.37410071942446044,
   "ppFgTeamAstPct": 0.09475178477907331,
   "teamAssistedEfg": 0.5065576186265842,
   "ptsProd": 569.54989314123,
   "ppOrb": 9.549382927325281,
   "ppAssist": 88.78978224455611,
   "ppAssist_Classic": 82.88469256002232,
   "ppFg": 299.2107279693486,
   "ftPoss": 100.69999999999999,
   "actualFtaToPoss": 0.451,
   "ftPct": 0.8113207547169812,
   "missedBothFTs": 0.03559985760056958,
   "offPlaysLessPoss": 74.21547619047624,
   "offPoss": 451.4845238095238,
   "fgPart": 125.8295019157088,
   "ftPart": 97.11509433962264,
   "astPart": 35.8574120603015,
   "astPart_Classic": 35.857412060301506,
   "orbPart": 4.4454424748097345,
   "teamScoringPoss": 901.7150259067357,
   "teamPlays": 2124.025,
   "adjPoss": 479.68306783293383,
   "scoringPoss": 263.2474507904427,
   "fgxPoss": 143.85071138211381,
   "ftxPoss": 3.5849056603773564,
   "oRtg": 118.73462528379162,
   "oRtg_Classic": 118.36218608211652,
   "defSos": 96.60184357541898,
   "avgEff": 100,
   "SD_at_Usage": 9.415132551133661,
   "SDs_Above_Mean": 1.9898419041945201,
   "SD_at_Usage_20": 10.143,
   "Regressed_ORtg": 120.18296643424502,
   "Usage": 25.0546350615718,
   "Usage_Bonus": 6.3182938269647515,
   "adjORtg": 130.9511864154515,
   "adjORtgPlus": 6.190237283090301
};
