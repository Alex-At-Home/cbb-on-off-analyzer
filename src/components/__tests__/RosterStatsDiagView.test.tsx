import renderer from 'react-test-renderer';
import React from 'react';
import RosterStatsDiagView from '../RosterStatsDiagView';
import { sampleOrtgDiagnostics } from "../../sample-data/sampleOrtgDiagnostics";

describe("RosterStatsDiagView", () => {
  test("RosterStatsDiagView - should create snapshot", () => {
    const testData = null;
    const component = renderer.create(<RosterStatsDiagView ortgDiags={sampleOrtgDiagnostics} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
