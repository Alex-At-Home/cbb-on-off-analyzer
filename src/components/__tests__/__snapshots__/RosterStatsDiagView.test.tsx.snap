// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RosterStatsDiagView RosterStatsDiagView - should create snapshot 1`] = `
<span>
  ORtg = Points_Produced/Adjusted_Possessions [
  <b>
    508.9
  </b>
  ] / [
  <b>
    450.3
  </b>
  ]  (
  <a
    href="#"
    onClick={[Function]}
  >
    more
     about points
  </a>
  )  (
  <a
    href="#"
    onClick={[Function]}
  >
    more
     about possessions
  </a>
  )
  <ul>
    <li>
      Points_Produced: [
      <b>
        508.9
      </b>
      ] = PProd_From_ORB [
      <b>
        9.5
      </b>
      ] + [
      <b>
        499.4
      </b>
      ] (PProd_From_FG [
      <b>
        301.5
      </b>
      ] + FTM [
      <b>
        166
      </b>
      ] + PProd_From_AST [
      <b>
        81.1
      </b>
      ], less Team_ORB_Contrib% [
      <b>
        9.0
        %
      </b>
      ]==[
      <b>
        49.2
      </b>
      ])
    </li>
    <ul>
      <li>
        <em>
          Compare raw stats: orb=[
          <b>
            17
          </b>
          ], pts=[
          <b>
            497
          </b>
          ] (fg_pts=[
          <b>
            331
          </b>
          ] + ftm=[
          <b>
            166
          </b>
          ]), assists=[
          <b>
            144
          </b>
          ]
        </em>
      </li>
    </ul>
    <li>
      Adjusted_Possessions: [
      <b>
        450.3
      </b>
      ] = Scoring_Possessions [
      <b>
        236.5
      </b>
      ] + Missed_FG_Possessions [
      <b>
        141.2
      </b>
      ] + Missed_FT_Possessions [
      <b>
        3.7
      </b>
      ] + TO [
      <b>
        69
      </b>
      ]
    </li>
    <ul>
      <li>
        <em>
          Compare raw stats: poss=[
          <b>
            444.6
          </b>
          ] (fga=[
          <b>
            351
          </b>
          ] + 0.475*fta=[
          <b>
            97.8
          </b>
          ] + to=[
          <b>
            69
          </b>
          ] - orb=[
          <b>
            73.2
          </b>
          ])
        </em>
      </li>
    </ul>
  </ul>
</span>
`;
